File: tw/src/main/java/pl/edu/agh/philosophers/Statistician.java
package pl.edu.agh.philosophers;

import io.vavr.Tuple2;
import pl.edu.agh.util.Utils;

import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

public class Statistician {
    private static Statistician instance;
    private Map<Integer, List<Double>> waitTime = new ConcurrentHashMap<>();

    public static Statistician getInstance() {
        if (instance == null) {
            instance = new Statistician();
        }
        return instance;
    }

    public void addWaitTime(int number, double measurement) {
        waitTime.computeIfAbsent(number, i -> new LinkedList<>());
        waitTime.get(number).add(measurement);
    }

    public List<Tuple2<Integer, Double>> waitTimeMean() {
        return waitTime.entrySet().stream()
                .map(entry -> new Tuple2<>(entry.getKey(), Utils.listMean(entry.getValue())))
                .collect(Collectors.toList());
    }
}

File: philosopers_pyt/statistician.py
from functools import reduce

wait_time = {0: [], 1: [], 2: [], 3: [], 4: []}


def add_wait_time(number, measurement):
    wait_time[number].append(measurement)


def wait_time_mean():
    return list(map(lambda t: (t[0], list_mean(t[1])), wait_time.items()))


def list_mean(values):
    return reduce(lambda a, b: a + b, values, 0) / len(values)


def print_measurements(measurements):
    for number, measurement in measurements:
        print('{} {}'.format(number, measurement))
